index.ts

```diff
@@ -0,0 +1,476 @@
+	weightkg: number;
+}
+
+// TODO: name collision :(
+class Species {
+	static forGen(gen: Generation): Promise<Species>;
+
+	getName(name: string): string;
+	get(name: string): Species | undefined;
+	getByID(id: ID): Species | undefined;
+	next(): [ID, Species] | true;
+}
+
+// stats.ts
+
+type StatName = 'hp' | 'atk' | 'def' | 'spa' | 'spd' | 'spe';
+type StatsTable<T = number> = {[stat in StatName]: T};
+type BoostName = Exclude<StatName, 'hp'> | 'accuracy' | 'evasion';
+type BoostsTable<T = number> = {[boost in BoostName]: T};
+
+export class Stats {
+	static calc(stat: Stat, base: number, iv: number, ev: number, level: number, gen?: Generation);
+	static calc(stat: Stat, base: number, iv: number, ev: number, level: number, nature: Nature, gen?: Generation);
+
+	static get(name: string): StatName | undefined;
+	static display(s: string, full: boolean = true): string;
+
+	static fill(p: <Partial<StatsTable<T>, val: T): StatsTable<T>;
+	static itod(iv: number): number;
+	static dtoi(dv: number): number;
+	static getHPDV(pivs: Partial<StatsTable>): number;
+}
+
+// statuses.ts
+
+type StatusName = 'par' | 'psn' | 'frz' | 'slp' | 'brn' | 'tox';
+
+// sets.ts
+
+// TODO: other than species/moves the rest can be optional?
+interface PokemonSet<T extends string = string> {
+	name: string;
+	species: T;
+	item: T;
+	ability: T;
+	moves: T[];
+	nature: T;
+	gender: GenderName;
+	evs: StatsTable;
+	ivs: StatsTable;
+	level: number;
+	shiny?: boolean;
+	happiness?: number;
+	hpType?: string;
+	pokeball?: T;
+}
+
+class Sets {
+	static forGen(gen: Generation): Promise<Teams>;
+
+	packSet(s: PokemonSet<ID  | string>): string;
+	unpackSet(s: string): PokemonSet<ID>;
+	exportSet(s: PokemonSet<ID | string>): string;
+	importSet(s: string): PokemonSet<ID>;
+}
+
+// teams.ts
+
+class Teams {
+	static forGen(gen: Generation): Promise<Teams>;
+
+	packTeam(s: PokemonSet<ID | string>[]): string;
+	unpackTeam(s: string): PokemonSet<ID>[];
+	exportTeam(s: PokemonSet<ID | string>[][]): string;
+	exportTeams(s: PokemonSet<ID | string>[]): string;
+	importTeam(s: string): PokemonSet<ID>[];
+	importTeams(s: string): PokemonSet<ID>[][];
+}
+
+// types.ts
+
+type TypeName = 'Normal' | 'Fighting' | 'Flying' | 'Poison' | 'Ground' | 'Rock' | 'Bug' | 'Ghost' | 'Steel' |
+	'Fire' | 'Water' | 'Grass' | 'Electric' | 'Psychic' | 'Ice' | 'Dragon' | 'Dark' | 'Fairy' | '???';
+
+type TypeEffectiveness = 0 | 1 | 2 | 3; // 'normal' | 'weakness' | 'resistance' | 'immunity';
+
+interface Type {
+	id: ID;
+	name: TypeName;
+	gen: Generation;
+	category?: MoveCategory;
+
+	HPIVs: Partial<StatsTable>;
+	HPDVs: Partial<StatsTable>;
+
+	damageTaken: {[type in TypeName?]: TypeEffectiveness};
+}
+
+class Types {
+	static forGen(gen: Generation): Promise<Types>;
+
+	get(name: string): Type | undefined;
+	getByID(id: ID): Type | undefined;
+	hiddenPower(pivs: Partial<StatsTable>): {type: TypeName, basePower: number}|undefined;
+	next(): [ID, Type] | true;
+}
+
+// dex.ts
+
+class Dex {
```

Comment:
basically a convenient 'God-object' - just do `dex = await Dex.forGen(...)` once and then be able to use all the features synchronously.